lib/internal/repl/history.js

```diff
@@ -17,168 +25,405 @@ let debug = require('internal/util/debuglog').debuglog('repl', (fn) => {
 
-  if (historyPath === '') {
-    repl._historyPrev = _replHistoryMessage;
-    return ready(null, repl);
+    this[kHistoryPath] = ReplHistory[kGetHistoryPath](options);
+    this[kContext] = context;
+    this[kTimer] = null;
+    this[kWriting] = false;
+    this[kPending] = false;
+    this[kRemoveHistoryDuplicates] = options.removeHistoryDuplicates || false;
+    this[kHistoryHandle] = null;
+    this[kIsFlushing] = false;
+    this[kSize] = options.historySize ?? context.historySize ?? kHistorySize;
+    this[kHistory] = options.history ?? [];
+    this[kIndex] = -1;
   }
 
-  if (!historyPath) {
-    try {
-      historyPath = path.join(os.homedir(), '.node_repl_history');
-    } catch (err) {
-      _writeToOutput(repl, '\nError: Could not get the home directory.\n' +
-        'REPL session history will not be persisted.\n');
+  initialize(onReadyCallback) {
+    // Empty string disables persistent history
+    if (this[kHistoryPath] === '') {
+      // Save a reference to the context's original _historyPrev
+      this.historyPrev = this[kContext]._historyPrev;
+      this[kContext]._historyPrev = this[kReplHistoryMessage].bind(this);
+      return onReadyCallback(null, this[kContext]);
+    }
+
+    const resolvedPath = this[kResolveHistoryPath]();
+    if (!resolvedPath) {
+      ReplHistory[kWriteToOutput](
+        this[kContext],
+        '\nError: Could not get the home directory.\n' +
+        'REPL session history will not be persisted.\n',
+      );
+
+      // Save a reference to the context's original _historyPrev
+      this.historyPrev = this[kContext]._historyPrev;
+      this[kContext]._historyPrev = this[kReplHistoryMessage].bind(this);
+      return onReadyCallback(null, this[kContext]);
+    }
+
+    if (!this[kHasWritePermission]()) {
+      ReplHistory[kWriteToOutput](
+        this[kContext],
+        '\nAccess to FileSystemWrite is restricted.\n' +
+        'REPL session history will not be persisted.\n',
+      );
+      return onReadyCallback(null, this[kContext]);
+    }
+
+    this[kContext].pause();
+
+    this[kInitializeHistory](onReadyCallback).catch((err) => {
+      this[kHandleHistoryInitError](err, onReadyCallback);
+    });
+  }
+
+  addHistory(isMultiline, lastCommandErrored) {
+    const line = this[kContext].line;
+
+    if (line.length === 0) return '';
+
+    // If the history is disabled then return the line
+    if (this[kSize] === 0) return line;
+
+    // If the trimmed line is empty then return the line
+    if (StringPrototypeTrim(line).length === 0) return line;
+
+    // This is necessary because each line would be saved in the history while creating
+    // a new multiline, and we don't want that.
+    if (isMultiline && this[kIndex] === -1) {
+      ArrayPrototypeShift(this[kHistory]);
+    } else if (lastCommandErrored) {
+      // If the last command errored and we are trying to edit the history to fix it
+      // Remove the broken one from the history
```

Comment:
```suggestion
      // remove the broken one from the history
```
(since this sentence is part of the multiline comment)

Reply:
This is not allowed by the linter